shared/route-tree/index.js

```diff
@@ -0,0 +1,229 @@
+      let newChild = _routeSet(childDef, childState, pathSpec.skip(1))
+      if (pathHead && pathHead.hasOwnProperty('props')) {
+        newChild = newChild.set('props', I.fromJS(pathHead.props))
+      }
+      return newChild
+    })
+  }
+
+  return newRouteState
+}
+
+export function routeSetProps (routeDef: RouteDefNode, routeState: ?RouteStateNode, pathProps: PathParam, parentPath: ?Path): RouteStateNode {
+  const pathSeq = I.Seq(pathProps).map(item => {
+    if (typeof item === 'string') {
+      return {type: 'navigate', next: item}
+    } else {
+      const {selected, ...props} = item
+      return {type: 'navigate', next: selected, props}
+    }
+  })
+  const parentPathSeq = I.Seq(parentPath || []).map(item => {
+    return {type: 'next', next: item}
+  })
+  return _routeSet(routeDef, routeState, parentPathSeq.concat(pathSeq))
+}
+
+export function routeNavigate (routeDef: RouteDefNode, routeState: ?RouteStateNode, pathProps: PathParam, parentPath: ?Path): RouteStateNode {
+  return routeSetProps(routeDef, routeState, I.List(pathProps).push({selected: null}), parentPath)
+}
+
+export function routeSetState (routeDef: RouteDefNode, routeState: RouteStateNode, path: Path, partialState: {}): RouteStateNode {
+  const pathSeq = I.Seq(path)
+  if (!pathSeq.size) {
+    return routeState.update('state', state => state.merge(partialState))
+  }
+  return routeState.updateChild(pathSeq.first(),
+    childState => routeSetState(routeDef, childState, pathSeq.skip(1), partialState)
+  )
+}
+
+export function routeClear (routeState: ?RouteStateNode, path: Path): ?RouteStateNode {
+  if (!routeState) {
+    return null
+  }
+  const pathSeq = I.Seq(path)
+  if (!pathSeq.size) {
+    return null
+  }
+  return routeState.updateChild(pathSeq.first(),
+    childState => routeClear(childState, pathSeq.skip(1))
+  )
+}
+
+// Traverse a routeState, making sure it matches the routeDef and ends on a leaf component.
+export function checkRouteState (routeDef: RouteDefNode, routeState: ?RouteStateNode): ?string {
+  if (!routeDef) {
+    return
+  }
+
+  let path = []
+  let curDef = routeDef
+  let curState = routeState
+  while (curState && curState.selected !== null) {
+    path.push(curState.selected)
+    curDef = curDef.getChild(curState.selected)
+    curState = curState.getChild(curState.selected)
+    if (!curDef) {
+      return `Missing route def: ${pathToString(path)}`
+    }
+  }
+  if (!curState) {
+    return `Route missing state: ${pathToString(path)}`
+  }
+  if (!curDef.component) {
+    return `Route missing component: ${pathToString(path)}`
+  }
+}
+
+export function getPath (routeState: RouteStateNode, parentPath?: Path): I.List<string> {
+  const path = []
+  let curState = routeState
+
+  if (parentPath) {
+    I.Seq(parentPath).forEach(next => {
+      curState = curState.getChild(next)
+      if (!curState) {
+        return path
+      }
+      path.push(next)
+    })
+  }
+
+  while (curState && curState.selected !== null) {
```

Comment:
Yes, `parentPath` specifies a starting point for the traversal, but we want to get the whole selected path under it. This lets us do:

(starting on /settings/invites/confirm)
`getPath(['settings']) => ['settings', 'invites', 'confirm']`

Or perhaps I didn't understand the question?